shell_automaton/src/mempool/mempool_state.rs

```diff
@@ -69,6 +69,71 @@ impl MempoolState {
             .get(&peer)
             .map_or(false, |p| p.seen_operations.contains(op_hash))
     }
+
+    /// Get next operation with highest priority for prevalidation.
+    pub fn next_for_prevalidation(&self) -> Option<(&OperationHash, &Operation)> {
+        self.injecting_rpc_ids
+            .iter()
+            .find_map(|(hash, _)| {
+                self.pending_operations
+                    .get(hash)
+                    .map(|content| (hash, content))
+            })
+            .or_else(|| self.pending_operations.next_for_prevalidation())
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Default, Clone)]
+pub struct MempoolPendingOperations {
+    ops: HashMap<OperationHash, Operation>,
+    consensus_ops: VecDeque<OperationHash>,
+}
+
+impl MempoolPendingOperations {
+    pub fn is_empty(&self) -> bool {
+        self.ops.is_empty()
+    }
+
+    pub fn contains_key(&self, key: &OperationHash) -> bool {
+        self.ops.contains_key(key)
+    }
+
+    pub fn get(&self, key: &OperationHash) -> Option<&Operation> {
+        self.ops.get(key)
+    }
+
+    pub fn insert(&mut self, key: OperationHash, value: Operation) {
+        let op_kind = OperationKind::from_operation_content_raw(value.data().as_ref());
+        if self.ops.insert(key.clone(), value).is_none() {
+            if op_kind.is_consensus_operation() {
+                self.consensus_ops.push_back(key);
+            }
+        }
+    }
+
+    pub fn remove(&mut self, key: &OperationHash) -> Option<Operation> {
+        if let Some(op_content) = self.ops.remove(key) {
+            if let Some(index) = self.consensus_ops.iter().position(|hash| hash == key) {
+                self.consensus_ops.remove(index);
+            }
+            Some(op_content)
+        } else {
+            None
+        }
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (&OperationHash, &Operation)> {
+        self.ops.iter()
+    }
+
+    /// Get next operation with highest priority for prevalidation.
+    pub fn next_for_prevalidation(&self) -> Option<(&OperationHash, &Operation)> {
```

Comment:
So this doesn't "pop" the operation right? it needs to be explicitly removed later for this call to returns something different (unless more operations are added), correct?

Reply:
Yes it doesn't pop. Only time we pop an operation is when operation was prevalidated or block was applied.